x/globalfee/ante/fee.go

```diff
@@ -0,0 +1,347 @@
+}
+
+func (mfd FeeDecorator) containsOnlyZeroFeeMsgs(ctx sdk.Context, msgs []sdk.Msg) (bool, error) {
+	for _, m := range msgs {
+		switch msg := m.(type) {
+		case *wasmtypes.MsgExecuteContract:
+			{
+				if !mfd.isZeroFeeMsg(ctx, msg) {
+					return false, nil
+				}
+			}
+		case *authz.MsgExec:
+			{
+				var authzMsgs []sdk.Msg
+				for _, v := range msg.Msgs {
+					var wrappedMsg sdk.Msg
+					err := mfd.codec.UnpackAny(v, &wrappedMsg)
+					if err != nil {
+						return false, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "error decoding authz messages")
+					}
+					authzMsgs = append(authzMsgs, wrappedMsg)
+				}
+				return mfd.containsOnlyZeroFeeMsgs(ctx, authzMsgs)
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func (mfd FeeDecorator) isZeroFeeMsg(ctx sdk.Context, msg *wasmtypes.MsgExecuteContract) bool {
+	contactAddr := sdk.MustAccAddressFromBech32(msg.Contract)
+	contractAuth, found := mfd.feeKeeper.GetContractAuthorization(ctx, contactAddr)
+	if found {
+		return isAuthorizedMethod(msg.GetMsg(), contractAuth.GetMethods())
+	}
+	codeId := mfd.feeKeeper.GetContractInfo(ctx, contactAddr).CodeID
+	codeAuth, found := mfd.feeKeeper.GetCodeAuthorization(ctx, codeId)
+	if found {
+		return isAuthorizedMethod(msg.GetMsg(), codeAuth.GetMethods())
+	}
+
+	return false
+}
+
+func isAuthorizedMethod(jsonBytes wasmtypes.RawContractMessage, methods []string) bool {
+	document := map[string]interface{}{}
+
+	if len(methods) == 1 && methods[0] == "*" {
+		return true
+	}
+
+	// contract method fetching taken from https://github.com/CosmWasm/wasmd/blob/4c906d5a53a255c551d6ed981a548cffe47ae9f0/x/wasm/types/json_matching.go
+	if err := json.Unmarshal(jsonBytes, &document); err != nil {
+		return false
+	}
+	if len(document) != 1 {
+		return false
+	}
+
+	for topLevelKey := range document {
+		for _, allowedKey := range methods {
+			if allowedKey == topLevelKey {
+				return true
+			}
+		}
+		return false
+	}
+	return false
+}
+
+// The fee checking ante mechanism below is based on the x/GlobalFee/ante from cosmos/gaia
+// https://github.com/cosmos/gaia/blob/6fe097e3280baa360a28b59a29b8cca964a5ae97/x/globalfee/ante/fee.go
+func (mfd FeeDecorator) checkFees(ctx sdk.Context, feeTx sdk.FeeTx, tx sdk.Tx, onlyZeroFeeMsgs bool, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	feeCoins := feeTx.GetFee().Sort()
+	gas := feeTx.GetGas()
+
+	// Get required Global Fee set by module
+	requiredGlobalFees, err := mfd.getGlobalFee(ctx, feeTx)
+	if err != nil {
+		return ctx, err
+	}
+
+	// Get local minimum-gas-prices set by the validator node
+	localFees := getMinGasPrice(ctx, int64(gas))
+
+	// CombinedFeeRequirement should never be empty since
+	// global fee is set to its default value, i.e. 0ustars, if empty
+	combinedFeeRequirement := combinedFeeRequirement(requiredGlobalFees, localFees)
```

Comment:
```suggestion
	combinedFeeRequirement := combinedFeeRequirement(requiredGlobalFees, localFees)
```
Change this variable name we are shadowing the method declared in Line 252